### DISCUSSION TITLE
AVL tree

### DISCUSSION THREAD

**Discussion by vumrra** [subak-ba/cs-study#42] [자료 구조]
# AVL tree

- “이진 탐색 트리”의 한 종류이며 “blance factor”를 통해 “스스로 균형을 유지"하는 트리이다.
- 트리의 모든 노드들은 blance factor가 -1 or 0 or 1 중 하나이다.
    - = 두 자식 서브트리의 높이는 최대 1만큰 차이난다.

### 이진트리

- 모든 노드가 자녀 노드를 최대 2개 가지는 트리

### 이진 트리 탐색 트리

- 이진 트리이며 이진 트리에 있는 모든 노드가 왼쪽 서브트리에는 자기보다 작은 값, 오른쪽 서브트리에는 자기보다 큰 값을 가지는 트리

## Blance factor

임의의 노드 n에 대해서

BF(n) = h(왼쪽서브트리(n)) - h(오른쪽서브트리(n))

*h: height 높이

*BF: blance factor

![image](https://github.com/subak-ba/cs-study/assets/131235625/73b52fd4-4880-465a-bba0-b796279b2db3)

> 예시 그림에서 100 노드에서 BF는 1 - 2 = -1이다.
> 

> 추가로 50 노드의 BF는 왼쪽 서브트리 높이 0 오른쪽 서브트리의 높이(노드가 없다) -1 이기 때문에 -1이다.
> 

> 리프 노드의 서브트리의 높이 차는 모두 0이기 때문에 BF = 0이다.
> 

### AVL 트리의 특징

트리의 모든 노드는 BF가 1 or 0 or 1 중 하나이다.

### AVL 트리가 균형을 잡는 과정

트리에 삽입, 삭제가 발생한다면

BF가 -1 or 0 or 1 이 아닌 노드가 생기면 균형을 맞추는 작업을 진행한다.

### 오른쪽 - 오른쪽으로 편향된 상태

![image](https://github.com/subak-ba/cs-study/assets/131235625/09d8f339-1fee-4b7b-802a-27905c1a805a)

해당 AVL 트리에서 200 값을 insert 한다면

![image](https://github.com/subak-ba/cs-study/assets/131235625/a064668e-b122-41cd-a780-92b566f3187d)

150 BF = -1

100 BF = -1 - 1 = **-2**

이렇게 BF 검사를 했을 때 BF가 -1 or 0 or 1 중 하나가 아닌 노드가 발생하게 된다.

이렇게 AVL가 100을 기준으로 오른쪽 - 오른쪽으로 편향된 상태이다.

- 100 - 150 - 200
- 150 노드가 중앙값이기 때문에 해당 노드가 가운데에 위치해야한다. 그렇기 때문에 해당 노드를 한 단계 위쪽으로 위치시켜 조건을 만족시킨다.

이런 상황에서는 150 노드를 위로 올려 (왼쪽으로 로테이트) AVL 트리의 조건을 만족시킨다.
![image](https://github.com/subak-ba/cs-study/assets/131235625/1dc0846a-98ca-45ec-b0d4-5526d2144176)

### 왼쪽 - 왼쪽으로 편향된 상태

![image](https://github.com/subak-ba/cs-study/assets/131235625/a4fb7e85-176f-4c4e-a5fe-a4a0e1060225)

해당 AVL 트리에서 30을 insert 한다면

![image](https://github.com/subak-ba/cs-study/assets/131235625/0a14127e-6e79-466f-a921-2ee4c2918944)

60 BF: -1

100 BF: 1

150 BF: 2

이렇게 BF 검사를 했을 때 BF가 -1 or 0 or 1 중 하나가 아닌 노드가 발생하게 된다.

이렇게 AVL가 150을 기준으로 왼쪽 - 왼쪽으로 편향된 상태이다.

이런 상태에서는 100 - 60 - 30 이렇게 세 노드가 편향을 이르키는 노드들이다. 해당 노드에 집중하여 편향을 해결하면 된다.

오른쪽 -오른쪽 편향과 똑같이 60 노드를 위로 한 단계 위로 올린다.

![image](https://github.com/subak-ba/cs-study/assets/131235625/9d5f074e-ac92-4af8-88ef-98ba2fce7851)

60 노드를 트리의 제일 위로 올렸을 때 100 노드가 60의 오른쪽 자녀가 2개의 노드를 가르키는 문제가 생긴다.

해당 문제는 150 노드가 100 노드를 왼쪽 서브트리 자리로 이동시켜주면 된다.

![image](https://github.com/subak-ba/cs-study/assets/131235625/f39adcb8-46b1-4fb3-8d92-986cfab8ac25)

이렇게 된다면 모든 노드가 이진 탐색트리, AVL 트리의 조건을 만족하게 된다.

### 오른쪽 - 왼쪽으로 편향된 상태

![image](https://github.com/subak-ba/cs-study/assets/131235625/a5546c8f-479d-4e5f-8e9d-b9dad799cd16)

해당 AVL 트리에서 1 노드를 delete 한다면

![image](https://github.com/subak-ba/cs-study/assets/131235625/3c822e1b-d819-4937-aa47-556a95a69aa4)

30 BF: -1

60 BF: -2

이렇게 BF 검사를 했을 때 BF가 -1 or 0 or 1 중 하나가 아닌 노드가 발생하게 된다.

이렇게 AVL가 60을 기준으로 오른쪽 - 왼쪽으로 편향된 상태이다.

이런 상태에서는 60 - 150 - 100 이렇게 세 노드가 편향을 이르키는 노드들이다. 해당 노드에 집중하여 편향을 해결하면 된다.

100 노드가 중앙 값이므로 100 노드를 60 노드 위로 올려서 해결하면 된다.

하지만 이 작업은 기존과 다르게 두 단계를 올려야 하므로, 구현에 관점에서는 두 단계를 통해서 해당 작업이 일어난다.

1. 100노드를 150노드 위로 올린다. (오른쪽으로 로테이트)

![image](https://github.com/subak-ba/cs-study/assets/131235625/ef844b19-d5cc-421f-b9cb-a2f65d0a7464)

1. 100 노드를 60 노드 위로 올린다. (왼쪽으로 로테이트)

![image](https://github.com/subak-ba/cs-study/assets/131235625/92a74da4-5355-4bc5-898d-9c9699b242d2)

이렇게 되면 AVL 트리 조건을 만족하게 된다.

### 왼쪽 - 오른쪽으로 편향된 상태

왼쪽 - 오른쪽으로 편향된 상태이다.

![image](https://github.com/subak-ba/cs-study/assets/131235625/3a8d0757-26de-449f-938d-582aa0a50fea)

60 BF: 2

이렇게 AVL가 60을 기준으로 왼쪽 - 오른쪽으로 편향된 상태이다.

이런 상태에서는 60 - 30- 50 이렇게 세 노드가 편향을 이르키는 노드들이다. 해당 노드에 집중하여 편향을 해결하면 된다.

50노드가 중앙 값이므로 50 노드를 60 노드 위로 올려서 해결하면 된다.

1. 50 노드를 30 노드 위로 올린다 (왼쪽으로 로테이트)

![image](https://github.com/subak-ba/cs-study/assets/131235625/28cc38e8-58f7-4a15-8c02-339a0757e4fe)

1. 50 노드를 60 노드 위로 올린다 (오른쪽으로 로테이트)

![image](https://github.com/subak-ba/cs-study/assets/131235625/82ad45b7-74d1-431e-9c72-92f235093529)

---

## 정리

- 삽입, 삭제는 일반적인 이진 탐색 트리와 동일하다
- 삽입, 삭제가 발생했을 때 발생한 위치에서 루트노드까지 올라오면서 해당 노드의 BF를 확인한다. 만약 균형이 깨졌다면 재조정을 진행한다.

## 장점

AVL 트리의 가장 큰 장점은 시간 복잡도이다.

일반적인 이진 탐색 트리와 best, avg case에서는 동일하게 삽입, 삭제, 조회 시간 복잡도가 각각 O(1), O(lon N) 이다.

하지만 worst case에서 이진 탐색 트리는 O(N)의 시간복잡도라면 AVL 트리는 엄격하게 균형을 맞춰주기 때문에 O(logN)으로 크게 개선이 된다.

## 단점

엄격하게 트리의 균형을 유지하기 때문에, 삽입, 삭제시 발생한 노드부터 루트 노드까지 올라가 BF를 확인하고 만약 균형이 깨졌다면 트리 구조를 재조정 하기 때문에 많은 시간이 소요된다.

→ 이를 개선한 트리가 Red-Black tree 이다.

**Comment by jangwooooo** [comment]
> 트리의 모든 노드는 BF가 1 or 0 or 1 중 하나이다.  

-1 오타 난 거 같아요 ~

**Comment by jangwooooo** [comment]
> 60 BF: -1
> 100 BF: 1  

왼쪽 - 왼쪽 편향 예제에서 해당 값들 잘못된 거 같아요.
60의 경우에는 BF가 1, 100의 경우에는 BF가 2 인 게 맞는 거 같아요.
제가 틀렸을 수도 있습니다 ㅎㅎ..

**Reply by jangwooooo** [reply]
> 30 BF: -1  

오른쪽 - 왼쪽 예제도 위와 같습니다

**Comment by jangwooooo** [comment]
> 장점
> worst case에서 이진 탐색 트리는 O(N)의 시간복잡도라면 AVL 트리는 엄격하게 균형을 맞춰주기 때문에 O(logN)으로 크게 개선이 된다.

> 단점
> 엄격하게 트리의 균형을 유지하기 때문에, 삽입, 삭제시 발생한 노드부터 루트 노드까지 올라가 BF를 확인하고 만약 균형이 깨졌다면 트리 구조를 재조정 하기 때문에 많은 시간이 소요된다.

두 번째 말대로 삽입, 삭제시 균형을 맞추기 위해 많은 시간이 필요할 거 같은데 그럼에도 AVL트리는 O(logN)의 시간복잡도를 가질 수 있나요?
